File: MobiTaskClient/app/src/main/java/com/task/mobi/fragment/ProfileFragment.java
package com.task.mobi.fragment;

import android.Manifest;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;
import com.task.mobi.R;
import com.task.mobi.activity.MainActivity;
import com.task.mobi.data.UserData;
import com.task.mobi.utils.AppUtils;
import com.task.mobi.utils.Const;
import com.task.mobi.utils.ImageUtils;
import com.task.mobi.utils.PermissionsManager;

import java.io.File;
import java.lang.reflect.Type;

import butterknife.BindView;
import butterknife.OnClick;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;

/**
 * Created by Neeraj Narwal on 5/5/17.
 */
public class ProfileFragment extends BaseFragment implements ImageUtils.ImageSelectCallback, PermissionsManager.PermissionCallback {

    public View view;

    @BindView(R.id.nameET)
    EditText nameET;

    @BindView(R.id.phoneET)
    EditText phoneET;

    @BindView(R.id.emailET)
    EditText emailET;

    @BindView(R.id.picIV)
    ImageView picIV;

    @BindView(R.id.updateBT)
    Button updateBT;

    @BindView(R.id.changePassBT)
    Button changePassBT;

    private Call<JsonObject> updateCall;
    private UserData userData;
    private boolean isEdit;
    private File imageFile;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (view == null) {
            view = inflater.inflate(R.layout.fg_profile, null);
        }
        setToolbar(true, "Profile");
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initUI();
    }

    private void initUI() {
        userData = (UserData) store.getUserData(Const.USER_DATA, UserData.class);
        nameET.setText(userData.name);
        phoneET.setText(userData.phone);
        emailET.setText(userData.email);
        AppUtils.MyTarget target = new AppUtils.MyTarget(picIV);
        baseActivity.picasso.load(Const.ASSETS_BASE_URL + userData.photo).error(R.mipmap.ic_default_user).placeholder(R.mipmap.ic_default_user).into(target);
        enableEdit(false);
    }

    private void enableEdit(boolean isEdit) {
        this.isEdit = isEdit;
        nameET.setFocusableInTouchMode(isEdit);
        nameET.setEnabled(isEdit);
        phoneET.setFocusableInTouchMode(isEdit);
        phoneET.setEnabled(isEdit);
        picIV.setOnClickListener(isEdit ? this : null);
        updateBT.setText(isEdit ? "Submit" : "Update Info");
        changePassBT.setVisibility(isEdit ? View.GONE : View.VISIBLE);
        if (isEdit) {
            nameET.requestFocus();
        }
    }

    @OnClick(R.id.updateBT)
    public void submit() {
        if (isEdit) {
            if (validate()) {
                updateProfile();
            }
        } else {
            enableEdit(true);
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.picIV:
                if (PermissionsManager.checkPermissions(baseActivity, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.CAMERA}, 122, this)) {
                    selectPic();
                }
                break;
        }
    }

    @OnClick(R.id.changePassBT)
    public void changePass() {
        baseActivity.getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.frame_container, new NewPasswordFragment())
                .addToBackStack(null)
                .commit();
    }

    private void selectPic() {
        new ImageUtils.ImageSelect.Builder(baseActivity, this, 123).aspectRatio(5, 4).start();
    }

    private void updateProfile() {
        RequestBody sessId = RequestBody.create(MediaType.parse("text/plain"), store.getString(Const.SESSION_ID));
        RequestBody name = RequestBody.create(MediaType.parse("text/plain"), getText(nameET));
        RequestBody phone = RequestBody.create(MediaType.parse("text/plain"), getText(phoneET));
        MultipartBody.Part image = null;
        try {
            image = MultipartBody.Part.createFormData("pic"
                    , imageFile.getName()
                    , RequestBody.create(MediaType.parse("image/jpeg"), imageFile));
        } catch (Exception e) {
            e.printStackTrace();
        }

        updateCall = apiInterface.updateProfile(sessId, name, phone, image);
        apiManager.makeApiCall(updateCall, this);
    }

    private boolean validate() {
        if (getText(nameET).isEmpty()) {
            showToast("Please enter Name", true);
        } else if (getText(phoneET).isEmpty()) {
            showToast("Please enter Phone Number", true);
        } else
            return true;
        return false;
    }

    @Override
    public void onSuccess(Call call, Object object) {
        super.onSuccess(call, object);
        if (updateCall == call) {
            JsonObject jsonObject = (JsonObject) object;
            JsonObject data = jsonObject.getAsJsonArray("data").get(0).getAsJsonObject();
            Type objectType = new TypeToken<UserData>() {
            }.getType();


            UserData userDataOld = (UserData) store.getUserData(Const.USER_DATA, UserData.class);
            String compId = userDataOld.cId;

            UserData userData = new Gson().fromJson(data, objectType);
            userData.cId = compId;
            store.saveString(Const.USER_ID, userData.id);
            store.saveUserData(Const.USER_DATA, userData);
            ((MainActivity) baseActivity).setUserData();
            showToast("Profile Updated Successfully");
            baseActivity.onBackPressed();
        }
    }

    @Override
    public void onImageSelected(String imagePath, int resultCode) {
        Bitmap bitmap = ImageUtils.imageCompress(imagePath);
        imageFile = ImageUtils.bitmapToFile(bitmap, baseActivity);
        picIV.setImageBitmap(bitmap);
        log("File Size >>>>>>>>." + imageFile.length());
    }

    @Override
    public void onPermissionsGranted(int resultCode) {
        selectPic();
    }

    @Override
    public void onPermissionsDenied(int resultCode) {

    }
}

File: MobiTaskAdmin/app/src/main/java/com/task/mobiadmin/fragment/ProfileFragment.kt
package com.task.mobiadmin.fragment

import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AlertDialog
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.android.gms.maps.model.LatLng
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.reflect.TypeToken
import com.task.mobiadmin.R
import com.task.mobiadmin.activity.MainActivity
import com.task.mobiadmin.data.UserData
import com.task.mobiadmin.utils.AppUtils
import com.task.mobiadmin.utils.Const
import com.task.mobiadmin.utils.Utils
import kotlinx.android.synthetic.main.fg_profile.*
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Call

/**
 * Created by Neeraj Narwal on 5/5/17.
 */
class ProfileFragment : BaseFragment(), PinAdd.AddPinListener {

    private lateinit var userData: UserData
    private var updateCall: Call<JsonObject>? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        setToolbar(false, "")
        return inflater!!.inflate(R.layout.fg_profile, null)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        baseActivity.setSupportActionBar(toolBarProfile)
        toolBarProfile.setNavigationOnClickListener { baseActivity.onBackPressed() }
        initUI()
    }

    private fun initUI() {
        userData = store.getUserData(Const.USER_DATA, UserData::class.java) as UserData
        activeTasksTV.text = (baseActivity as MainActivity).getTotalTasksCount() + " Active tasks"
        val target = AppUtils.MyTarget(userIV, baseActivity, false)
        baseActivity.picasso.load(Const.ASSETS_BASE_URL + userData.photo).placeholder(R.mipmap.ic_default_company).error(R.mipmap.ic_default_company).into(target)
        userIV.tag = target
        toolBarProfile.title = userData.companyname
        emailTV.text = userData.email
        phoneTV.text = userData.phoneno
        addressTV.text = userData.address
        expiryDateTV.text = "Active until : " + Utils.changeDateFormat(userData.expiryTime, "yyyy-MM-dd HH:mm:ss", "MMM dd, yyyy")
        if (userData.lat.isEmpty()) {
            addLocCL.visibility = View.VISIBLE
            addLocCL.setOnClickListener {
                val pinAdd: PinAdd = PinAdd.instance
                pinAdd.setListener(this@ProfileFragment)
                baseActivity.supportFragmentManager
                        .beginTransaction()
                        .replace(R.id.frame_container, AddLocationPinFragment())
                        .addToBackStack(null)
                        .commit()
            }
        } else AppUtils().loadStaticMap(baseActivity, locIV, userData.lat, userData.lng, true)

        editIV.setOnClickListener {
            baseActivity.supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.frame_container, ProfileUpdateFragment())
                    .addToBackStack(null)
                    .commit()
        }
        saveBT.setOnClickListener { updateDetails() }
        passBT.setOnClickListener {
            baseActivity
                    .supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.frame_container, NewPasswordFragment())
                    .addToBackStack(null)
                    .commit()
        }
    }

    private fun updateDetails() {
        val sessionId = RequestBody.create(MediaType.parse("text/plain"), store.getString(Const.SESSION_ID))
        val name = RequestBody.create(MediaType.parse("text/plain"), userData.companyname)
        val phone = RequestBody.create(MediaType.parse("text/plain"), userData.phoneno)
        val lat = RequestBody.create(MediaType.parse(" text/plain"), userData.lat)
        val lng = RequestBody.create(MediaType.parse(" text/plain"), userData.lng)
        val address = RequestBody.create(MediaType.parse("text/plain"), userData.address)
        val image: MultipartBody.Part? = null
        updateCall = apiInterface.updateProfile(sessionId, name, phone, lat, lng, address, image)
        apiManager.makeApiCall(updateCall, this)
    }

    override fun onLocationPinAdded(address: String, latLng: LatLng) {
        Handler().postDelayed({
            baseActivity.runOnUiThread {
                addLocCL.visibility = View.GONE
                saveBT.visibility = View.VISIBLE
                AppUtils().loadStaticMap(baseActivity, locIV, latLng.latitude.toString(), latLng.longitude.toString(), true)
                userData.lat = latLng.latitude.toString()
                userData.lng = latLng.longitude.toString()
                showAddressDialog(address)
            }
        }, 400)
    }

    private fun showAddressDialog(address: String) {
        val bldr = AlertDialog.Builder(baseActivity)
        bldr.setTitle("Address")
        bldr.setMessage("Do you also want to update address \n\n${userData.address}\n\n with \n\n$address\n")
        bldr.setPositiveButton("Update") { _, _ ->
            userData.address = address
            addressTV.text = userData.address
        }
        bldr.setCancelable(false)
        bldr.setNegativeButton("No", null)
        bldr.create().show()
    }

    override fun onSuccess(call: Call<*>?, `object`: Any?) {
        super.onSuccess(call, `object`)
        if (updateCall === call) {

            val jsonObject = `object` as JsonObject
            val data = jsonObject.getAsJsonObject("data")
            val objectType = object : TypeToken<UserData>() {}.type
            val userData = Gson().fromJson<UserData>(data, objectType)
            userData.sessionId = store.getString(Const.SESSION_ID)
            store.saveString(Const.USER_ID, userData.id)
            store.saveUserData(Const.USER_DATA, userData)

            (baseActivity as MainActivity).setUserData()

            showToast("Profile updated successfully")
            baseActivity.onBackPressed()
        }
    }
}
